com.unity.multiplayer.mlapi/Runtime/Core/NetworkBehaviour.cs

```diff
@@ -46,34 +46,44 @@ internal NetworkSerializer __beginSendServerRpc(uint rpcMethodId, ServerRpcParam
         {
             PooledNetworkWriter writer;
 
-            var rpcQueueContainer = NetworkManager.RpcQueueContainer;
-            var isUsingBatching = rpcQueueContainer.IsUsingBatching();
-            var transportChannel = rpcDelivery == RpcDelivery.Reliable ? NetworkChannel.ReliableRpc : NetworkChannel.UnreliableRpc;
-
-            if (IsHost)
+            if (serverRpcParams.Send.UpdateStage == NetworkUpdateStage.Unset)
             {
-                writer = rpcQueueContainer.BeginAddQueueItemToFrame(RpcQueueContainer.QueueItemType.ServerRpc, Time.realtimeSinceStartup, transportChannel,
-                    NetworkManager.ServerClientId, null, RpcQueueHistoryFrame.QueueFrameType.Inbound, serverRpcParams.Send.UpdateStage);
+                serverRpcParams.Send.UpdateStage = NetworkUpdateLoop.UpdateStage;
 
-                if (!isUsingBatching)
+                if (serverRpcParams.Send.UpdateStage == NetworkUpdateStage.Initialization)
                 {
-                    writer.WriteByte(NetworkConstants.SERVER_RPC); // MessageType
+                    serverRpcParams.Send.UpdateStage = NetworkUpdateStage.EarlyUpdate;
                 }
             }
+
+            if (serverRpcParams.Send.UpdateStage == NetworkUpdateStage.Initialization)
+            {
+                throw new NotSupportedException(
+                    $"{nameof(NetworkUpdateStage.Initialization)} cannot be used as a target for processing RPCs.");
+            }
```

Comment:
This feels more like an invariant rather than a normal operational condition - is assertion reasonable here?

Reply:
I had proposed using a separate enum for RPCs to prevent the user from being able to specify Initialization in the first place... the proposal was to create one for RPCs that was limited to just FixedUpdate, Update, and LateUpdate. I decided that I didn't want to bite that off right now because of the complication of trying to figure out how to handle Unset if this is called somehow in a stage other than one of those three, so I just added the exception for now, but I agree it's not a great solution - it's a temporary one with a better one to come when I figure out the details of the better one.

Reply:
Oh, I misread your comment, actually. Yes, I could assert instead of throwing, but I disagree it's an invariant since the user can specify the update stage on their own so this is actually a response to an invalid input to this function. I'm not sure if exception or assertion is correct there.

Reply:
Was this something we were going to abstract away from the user?
If so, then I think for the time being an exception is just fine.  However, we should make sure this scenario becomes "impossible" for a user.  I know this came up during your presentation, but remind me what the action item was to handle this was again?  I seem to remember a discussion revolving around not exposing this playerloop update stage to the users.

Reply:
Yes, I do think we should abstract it away, but when I started working on doing that, it spiraled into some other questions I didn't want to try to tackle with this work. For example - if we limit it to only FixedUpdate, Update, and LateUpdate like we discussed before, what happens if the user somehow initiates an RPC during a different stage? The behavior of "Unset" would have to get a lot more complex to properly shift it to the correct stage.

Since we're talking about possibly getting rid of the update stage field with the ticks system anyway, I wanted to wait and see how that shakes out before trying to tackle that.

Reply:
I think I'm fine with this exception, it does make sense.
IIRC, we discussed how we'd use a different enum other than `NetworkUpdateLoop`'s that is designed with users in mind. we might collab on this w @LukeStampfli and his upcoming `NetworkTickSystem` giving us a little bit more context and insights for the future.
(for the record, UpdateState enum never designed for the users in mind and we internally discussed making NetworkUpdateLoop -and its related types- completely internal when NetworkTickSystem & NetworkTimeSystem fully lands — that'd definitely have knock-on effects here too. let's keep all these as is -as far as this PR's concerned- but we'll modify stages and enum again in the future for sure)